File: src/main/java/com/dh/ondot/schedule/infra/dto/OdsayRouteApiResponse.java
package com.dh.ondot.schedule.infra.dto;

import java.util.Collections;
import java.util.List;

public record OdsayRouteApiResponse(
        Result result
) {
    public record Result(
            int searchType,
            int outTrafficCheck,
            int busCount,
            int subwayCount,
            int subwayBusCount,
            double pointDistance,
            int startRadius,
            int endRadius,
            List<Path> path
    ) {}

    public record Path(
            int pathType,
            Info info,
            List<SubPath> subPath
    ) {}

    public record Info(
            double trafficDistance,
            int totalWalk,
            int totalTime,
            int payment,
            int busTransitCount,
            int subwayTransitCount,
            String mapObj,
            String firstStartStation,
            String firstStartStationKor,
            String firstStartStationJpnKata,
            String lastEndStation,
            String lastEndStationKor,
            String lastEndStationJpnKata,
            int totalStationCount,
            int busStationCount,
            int subwayStationCount,
            double totalDistance,
            int totalWalkTime,
            int checkIntervalTime,
            String checkIntervalTimeOverYn,
            int totalIntervalTime
    ) {}

    public record SubPath(
            int trafficType,
            double distance,
            int sectionTime,
            Integer stationCount,
            List<Lane> lane,
            Integer intervalTime,
            String startName,
            String startNameKor,
            String startNameJpnKata,
            double startX,
            double startY,
            String endName,
            String endNameKor,
            String endNameJpnKata,
            double endX,
            double endY,
            String way,
            Integer wayCode,
            String door,
            Integer startID,
            Integer endID,
            Integer startStationCityCode,
            Integer startStationProviderCode,
            String startLocalStationID,
            String startArsID,
            Integer endStationCityCode,
            Integer endStationProviderCode,
            String endLocalStationID,
            String endArsID,
            String startExitNo,
            Double startExitX,
            Double startExitY,
            String endExitNo,
            Double endExitX,
            Double endExitY,
            PassStopList passStopList
    ) {}

    public record Lane(
            String name,
            String nameKor,
            String nameJpnKata,
            Integer subwayCode,
            Integer subwayCityCode,
            String busNo,
            String busNoKor,
            String busNoJpnKata,
            Integer type,
            Integer busID,
            String busLocalBlID,
            Integer busCityCode,
            Integer busProviderCode
    ) {}

    public record PassStopList(List<Station> stations) {}

    public record Station(
            int index,
            int stationID,
            String stationName,
            String stationNameKor,
            String stationNameJpnKata,
            Integer stationCityCode,
            Integer stationProviderCode,
            String localStationID,
            String arsID,
            String x,
            String y,
            String isNonStop
    ) {}

    public static OdsayRouteApiResponse walkOnly(int estimatedMinutes) {
        Info walkInfo = new Info(
                0.0, // trafficDistance
                0, // totalWalk
                estimatedMinutes,
                0, // payment
                0, 0,
                null, null, null, null, null, null, null,
                0, 0, 0,
                0.0, // totalDistance
                estimatedMinutes,
                0, null,
                0
        );

        Path walkPath = new Path(
                10, // custom code to indicate walk-only
                walkInfo,
                Collections.emptyList()
        );

        Result result = new Result(
                0, 0, 0, 0, 0, 0.0, 0, 0,
                List.of(walkPath)
        );

        return new OdsayRouteApiResponse(result);
    }
}

File: src/main/kotlin/com/dh/ondot/schedule/infra/dto/OdsayRouteApiResponse.kt
package com.dh.ondot.schedule.infra.dto

data class OdsayRouteApiResponse(
    val result: Result?,
) {
    data class Result(
        val searchType: Int,
        val outTrafficCheck: Int,
        val busCount: Int,
        val subwayCount: Int,
        val subwayBusCount: Int,
        val pointDistance: Double,
        val startRadius: Int,
        val endRadius: Int,
        val path: List<Path>?,
    )

    data class Path(
        val pathType: Int,
        val info: Info,
        val subPath: List<SubPath>,
    )

    data class Info(
        val trafficDistance: Double,
        val totalWalk: Int,
        val totalTime: Int,
        val payment: Int,
        val busTransitCount: Int,
        val subwayTransitCount: Int,
        val mapObj: String?,
        val firstStartStation: String?,
        val firstStartStationKor: String?,
        val firstStartStationJpnKata: String?,
        val lastEndStation: String?,
        val lastEndStationKor: String?,
        val lastEndStationJpnKata: String?,
        val totalStationCount: Int,
        val busStationCount: Int,
        val subwayStationCount: Int,
        val totalDistance: Double,
        val totalWalkTime: Int,
        val checkIntervalTime: Int,
        val checkIntervalTimeOverYn: String?,
        val totalIntervalTime: Int,
    )

    data class SubPath(
        val trafficType: Int,
        val distance: Double,
        val sectionTime: Int,
        val stationCount: Int?,
        val lane: List<Lane>?,
        val intervalTime: Int?,
        val startName: String?,
        val startNameKor: String?,
        val startNameJpnKata: String?,
        val startX: Double,
        val startY: Double,
        val endName: String?,
        val endNameKor: String?,
        val endNameJpnKata: String?,
        val endX: Double,
        val endY: Double,
        val way: String?,
        val wayCode: Int?,
        val door: String?,
        val startID: Int?,
        val endID: Int?,
        val startStationCityCode: Int?,
        val startStationProviderCode: Int?,
        val startLocalStationID: String?,
        val startArsID: String?,
        val endStationCityCode: Int?,
        val endStationProviderCode: Int?,
        val endLocalStationID: String?,
        val endArsID: String?,
        val startExitNo: String?,
        val startExitX: Double?,
        val startExitY: Double?,
        val endExitNo: String?,
        val endExitX: Double?,
        val endExitY: Double?,
        val passStopList: PassStopList?,
    )

    data class Lane(
        val name: String?,
        val nameKor: String?,
        val nameJpnKata: String?,
        val subwayCode: Int?,
        val subwayCityCode: Int?,
        val busNo: String?,
        val busNoKor: String?,
        val busNoJpnKata: String?,
        val type: Int?,
        val busID: Int?,
        val busLocalBlID: String?,
        val busCityCode: Int?,
        val busProviderCode: Int?,
    )

    data class PassStopList(val stations: List<Station>?)

    data class Station(
        val index: Int,
        val stationID: Int,
        val stationName: String?,
        val stationNameKor: String?,
        val stationNameJpnKata: String?,
        val stationCityCode: Int?,
        val stationProviderCode: Int?,
        val localStationID: String?,
        val arsID: String?,
        val x: String?,
        val y: String?,
        val isNonStop: String?,
    )

    companion object {
        fun walkOnly(estimatedMinutes: Int): OdsayRouteApiResponse {
            val walkInfo = Info(
                trafficDistance = 0.0,
                totalWalk = 0,
                totalTime = estimatedMinutes,
                payment = 0,
                busTransitCount = 0,
                subwayTransitCount = 0,
                mapObj = null,
                firstStartStation = null,
                firstStartStationKor = null,
                firstStartStationJpnKata = null,
                lastEndStation = null,
                lastEndStationKor = null,
                lastEndStationJpnKata = null,
                totalStationCount = 0,
                busStationCount = 0,
                subwayStationCount = 0,
                totalDistance = 0.0,
                totalWalkTime = estimatedMinutes,
                checkIntervalTime = 0,
                checkIntervalTimeOverYn = null,
                totalIntervalTime = 0,
            )

            val walkPath = Path(
                pathType = 10, // custom code to indicate walk-only
                info = walkInfo,
                subPath = emptyList(),
            )

            val result = Result(
                searchType = 0,
                outTrafficCheck = 0,
                busCount = 0,
                subwayCount = 0,
                subwayBusCount = 0,
                pointDistance = 0.0,
                startRadius = 0,
                endRadius = 0,
                path = listOf(walkPath),
            )

            return OdsayRouteApiResponse(result)
        }
    }
}
